File: TUnit.TestProject/DependsOnTests3.cs
using TUnit.TestProject.Attributes;

namespace TUnit.TestProject;

[EngineTest(ExpectedResult.Pass)]
public class DependsOnTests3
{
    private static DateTimeOffset _test3Start;

    private static DateTimeOffset _test1End;
    private static DateTimeOffset _test2End;

    [Test]
    public async Task Test1(CancellationToken cancellationToken)
    {
        var timeProvider = TimeProviderContext.Current;
        await timeProvider.Delay(TimeSpan.FromSeconds(1), cancellationToken);

        TestContext.Current!.StateBag.Items["Test1"] = "1";

        _test1End = timeProvider.GetUtcNow();
    }

    [Test]
    public async Task Test2(CancellationToken cancellationToken)
    {
        var timeProvider = TimeProviderContext.Current;
        await timeProvider.Delay(TimeSpan.FromSeconds(1), cancellationToken);

        TestContext.Current!.StateBag.Items["Test2"] = "2";

        _test2End = timeProvider.GetUtcNow();
    }

    [Test]
    [DependsOn(nameof(Test1))]
    [DependsOn(nameof(Test2))]
    public async Task Test3(CancellationToken cancellationToken)
    {
        var timeProvider = TimeProviderContext.Current;
        _test3Start = timeProvider.GetUtcNow();

        await timeProvider.Delay(TimeSpan.FromSeconds(1), cancellationToken);

        var test1 = TestContext.Current!.Dependencies.GetTests(nameof(Test1));
        var test2 = TestContext.Current.Dependencies.GetTests(nameof(Test2));

        await Assert.That(test1).HasCount().EqualTo(1);
        await Assert.That(test2).HasCount().EqualTo(1);

        await Assert.That(test1[0].StateBag.Items).ContainsKey("Test1");
        await Assert.That(test2[0].StateBag.Items).ContainsKey("Test2");
    }

    [After(Class)]
    public static async Task AssertStartTimes()
    {
        await Assert.That(_test3Start).IsAfterOrEqualTo(_test1End);
        await Assert.That(_test3Start).IsAfterOrEqualTo(_test2End);
    }
}

File: TUnit.TestProject.FSharp/DependsOnTests3.fs
namespace TUnit.TestProject.FSharp

open System
open System.Threading.Tasks
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Core

// Equivalent of DependsOnTests3.cs

type DependsOnTests3() =
    static let mutable test1Start = DateTime.MinValue
    static let mutable test2Start = DateTime.MinValue
    static let mutable test3Start = DateTime.MinValue

    [<Test>]
    member _.Test1() : Task = task {
        test1Start <- TestContext.Current.Value.TestStart.Value.DateTime
        do! Task.Delay(TimeSpan.FromSeconds(1.0))
        TestContext.Current.Value.ObjectBag.Add("Test1", box "1")
    }

    [<Test>]
    member _.Test2() : Task = task {
        test2Start <- TestContext.Current.Value.TestStart.Value.DateTime
        do! Task.Delay(TimeSpan.FromSeconds(1.0))
        TestContext.Current.Value.ObjectBag.Add("Test2", box "2")
    }

    [<Test>]
    [<DependsOn("Test1")>]
    [<DependsOn("Test2")>]
    member _.Test3() : Task = task {
        test3Start <- TestContext.Current.Value.TestStart.Value.DateTime
        do! Task.Delay(TimeSpan.FromSeconds(1.0))
        let test1 = TestContext.Current.GetTests("Test1")
        let test2 = TestContext.Current.GetTests("Test2")
        do! Assert.That(test1).HasCount().EqualTo(1)
        do! Assert.That(test2).HasCount().EqualTo(1)
        do! Assert.That(test1.[0].ObjectBag).ContainsKey("Test1")
        do! Assert.That(test2.[0].ObjectBag).ContainsKey("Test2")
    }

    [<After(HookType.Class)>]
    static member AssertStartTimes() : Task = task {
        do! Assert.That(test3Start).IsAfterOrEqualTo(test1Start.AddSeconds(0.9))
        do! Assert.That(test3Start).IsAfterOrEqualTo(test2Start.AddSeconds(0.9))
    }
